core/src/main/kotlin/com/exactpro/th2/read/db/bootstrap/Main.kt

```diff
@@ -297,7 +321,59 @@ private fun configureMessageStoring(
     return messagesQueue
 }
 
-private data class SessionKey(val alias: String, val direction: Direction)
+private val nanosInSecond = TimeUnit.SECONDS.toNanos(1)
+
+private fun protoKeyExtractor(builder: ProtoRawMessage.Builder): SessionKey<ProtoDirection> =
+    SessionKey(builder.sessionAlias, builder.direction)
+
+private fun transportKeyExtractor(builder: RawMessage.Builder): SessionKey<Direction> =
+    SessionKey(builder.idBuilder().sessionAlias, builder.idBuilder().direction)
+
+private abstract class Preprocessor<BUILDER, DIRECTION>(
+    protected val configBookName: String,
+    protected val configSessionGroup: String
+) {
+    protected val sequences = ConcurrentHashMap<SessionKey<DIRECTION>, Long>()
+    abstract fun preprocess(key: SessionKey<DIRECTION>, builder: BUILDER): BUILDER
+}
+
+private class ProtoPreprocessor(bookName: String, sessionGroup: String) : Preprocessor<ProtoRawMessage.Builder, ProtoDirection>(bookName, sessionGroup) {
+    override fun preprocess(key: SessionKey<ProtoDirection>, builder: ProtoRawMessage.Builder): ProtoRawMessage.Builder =
+        builder.apply {
+            sequence = sequences.compute(key) { _, prev ->
+                if (prev == null) {
+                    Instant.now().run { epochSecond * nanosInSecond + nano }
+                } else {
+                    prev + 1
+                }
+            }.let(::requireNotNull)
+            metadataBuilder.idBuilder.apply {
+                timestamp = Instant.now().toTimestamp()
+                sessionGroup = configSessionGroup
+                bookName = configBookName
+            }
+        }
+}
+
+private class TransportPreprocessor(bookName: String, sessionGroup: String) : Preprocessor<RawMessage.Builder, Direction>(bookName, sessionGroup) {
+    override fun preprocess(key: SessionKey<Direction>, builder: RawMessage.Builder): RawMessage.Builder {
+        builder.idBuilder().apply {
+            setSequence(requireNotNull(
+                sequences.compute(key) { _, prev ->
+                    if (prev == null) {
+                        Instant.now().run { epochSecond * nanosInSecond + nano }
+                    } else {
+                        prev + 1
+                    }
+                }
+            ))
+            setTimestamp(Instant.now())
+        }
+        return builder
+    }
+}
+
+private data class SessionKey<DIRECTION>(val alias: String, val direction: DIRECTION)
```

Comment:
I we use groups here the group must be used for the session key instead of the alias